src/main/java/edu/bu/metcs673/trackr/transaction/TransactionServiceImpl.java

```diff
@@ -22,106 +26,139 @@
+        if (transaction != null) {
+            return transaction;
+        }
+        throw new TrackrInputValidationException(CommonConstants.INVALID_TRANSACTION_ID);
+    }
+
+    /**
+     * The purpose of this method is to modify a special transaction record by
+     * 'transaction' and 'transactionInput' value.
+     *
+     * @param transaction      this is a Transaction object
+     * @param transactionInput This is a TransactionDTO object
+     * @author Xiaobing Hou
+     * @date 05/23/2022
+     */
+    @Override
+    public Transaction modifyTransaction(Transaction transaction, TransactionDTO transactionInput) {
+        transaction.setMoney(transactionInput.getMoney());
+        transaction.setTime(transactionInput.getTime());
+        transaction.setCounterparty(transactionInput.getCounterparty());
+        transaction.setTransactionDescription(transactionInput.getTransactionDescription());
+        transactionRepository.save(transaction);
+        return transaction;
+    }
+
+    /**
+     * The purpose of this method is to fina all the transaction of a bank account
+     *
+     * @param bankAccountId this is bank account id
+     * @return Transaction
+     * @author Xiaobing Hou
+     * @date 05/23/2022
+     */
+    @Override
+    public List<Transaction> findAllTraByBankAccountId(long bankAccountId) {
+        return transactionRepository.findAllTraByBankAccountIdAndStatus(bankAccountId,
+                Transaction.TRANSACTION_STATUS.VALID);
+    }
+
+    /**
+     * The purpose of this method is to invalid a transaction
+     *
+     * @param transaction this is a Transaction object
+     * @return Transaction
+     * @author Xiaobing Hou
+     * @date 05/25/2022
+     */
+    @Override
+    public Transaction deleteTransaction(Transaction transaction) {
+        transaction.setStatus(Transaction.TRANSACTION_STATUS.INVALID);
+        transactionRepository.save(transaction);
+        return transaction;
+    }
+
+    /**
+     * The purpose of this method is to get all valid transaction
+     *
+     * @param userId
+     * @return List<Transaction>
+     * @author Tim and Xiaobing
+     */
+    @Override
+    public List<Transaction> findAllTransactions(long userId) {
+
+
+        List<Long> bankAccountIds = bankAccountRepository.findAllByUserIdAndStatus(userId, ACCOUNT_STATUS.ACTIVE)
+                .stream().map(BankAccount::getId).collect(Collectors.toList());
+
+        List<Transaction> transactions = transactionRepository.findByBankAccountIdInAndStatus(bankAccountIds, Transaction.TRANSACTION_STATUS.VALID);
+        transactions.sort((t1, t2) -> {
+            long dateDifference = getMillisecond(t1.getTime()) - getMillisecond(t2.getTime());
+            if (dateDifference == 0) {
+                return 0;
+            } else {
+                return dateDifference > 0 ? -1 : 1;
+            }
+        });
```

Comment:
Order By Time DESC in the JPA query is a better approach.
```java
findByBankAccountIdInAndStatusOrderByTimeDesc()
```

Reply:
OK, have done
thanks so much!